drivers/serial/uart_nrf5e.c

```diff
@@ -0,0 +1,873 @@
+	__O u32_t  TASKS_STOPRX;
+	__O u32_t  TASKS_STARTTX;
+	__O u32_t  TASKS_STOPTX;
+
+	__I u32_t  RESERVED0[7];
+	__O u32_t  TASKS_FLUSHRX;
+
+	__I u32_t  RESERVED1[52];
+	__IO u32_t EVENTS_CTS;
+	__IO u32_t EVENTS_NCTS;
+	__IO u32_t EVENTS_RXDRDY;
+
+	__I u32_t  RESERVED2[1];
+	__IO u32_t EVENTS_ENDRX;
+
+	__I u32_t  RESERVED3[2];
+	__IO u32_t EVENTS_TXDRDY;
+	__IO u32_t EVENTS_ENDTX;
+	__IO u32_t EVENTS_ERROR;
+
+	__I u32_t  RESERVED4[7];
+	__IO u32_t EVENTS_RXTO;
+
+	__I u32_t  RESERVED5[1];
+	__IO u32_t EVENTS_RXSTARTED;
+	__IO u32_t EVENTS_TXSTARTED;
+
+	__I u32_t  RESERVED6[1];
+	__IO u32_t EVENTS_TXSTOPPED;
+
+	__I u32_t  RESERVED7[41];
+	__IO u32_t SHORTS;
+
+	__I u32_t  RESERVED8[63];
+	__IO u32_t INTEN;
+	__IO u32_t INTENSET;
+	__IO u32_t INTENCLR;
+
+	__I u32_t  RESERVED9[93];
+	__IO u32_t ERRORSRC;
+
+	__I u32_t  RESERVED10[31];
+	__IO u32_t ENABLE;
+
+	__I u32_t  RESERVED11[1];
+	__IO u32_t PSELRTS;
+	__IO u32_t PSELTXD;
+	__IO u32_t PSELCTS;
+	__IO u32_t PSELRXD;
+
+	__I u32_t  RESERVED12[3];
+	__IO u32_t BAUDRATE;
+
+	__I u32_t  RESERVED13[3];
+	__IO u32_t RXD;
+	__IO u32_t RXDMAXCNT;
+	__IO u32_t RXDAMOUNT;
+
+	__I u32_t  RESERVED14[1];
+	__IO u32_t TXD;
+	__IO u32_t TXDMAXCNT;
+	__IO u32_t TXDAMOUNT;
+
+	__I u32_t  RESERVED15[7];
+	__IO u32_t CONFIG;
+};
+
+/* Device data structure */
+struct uart_nrf5e_dev_data_t {
+	u32_t baud_rate;		/**< Baud rate */
+
+#ifdef CONFIG_UART_INTERRUPT_DRIVEN
+	uart_irq_callback_t  cb;  /**< Callback function pointer */
+	u8_t fifo_tx_data[CONFIG_UART_NRF5E_TX_DMA_BUFSIZE];
+	u8_t fifo_tx_read_index;
+	u8_t fifo_tx_write_index;
+	volatile u8_t fifo_tx_operating;
+
+	u8_t fifo_rx_data[CONFIG_UART_NRF5E_RX_DMA_BUFSIZE];
+	u8_t fifo_rx_read_index;
+	u8_t fifo_rx_write_index;
+	volatile u8_t fifo_rx_operating;
+
+	u8_t rx_irq_enable;
+	u8_t tx_irq_enable;
+
+	volatile u8_t in_interrupt;
+#endif /* CONFIG_UART_INTERRUPT_DRIVEN */
+};
+
+/* convenience defines */
+#define DEV_CFG(dev) \
+	((const struct uart_device_config * const)(dev)->config->config_info)
+#define DEV_DATA(dev) \
+	((struct uart_nrf5e_dev_data_t * const)(dev)->driver_data)
+#define UART_STRUCT(dev) \
+	((volatile struct _uart *)(DEV_CFG(dev))->base)
+
+#define UART_IRQ_MASK_ENDRX	(1 << 4)
+#define UART_IRQ_MASK_ENDTX	(1 << 8)
+#define UART_IRQ_MASK_ERROR	(1 << 9)
+
+static const struct uart_driver_api uart_nrf5e_driver_api;
+
+#ifdef CONFIG_UART_INTERRUPT_DRIVEN
+static int uart_nrf5e_irq_tx_ready(struct device *dev);
+static int uart_nrf5e_fifo_fill(struct device *dev,
+			 const u8_t *tx_data, int len);
+static int uart_nrf5e_fifo_read(struct device *dev,
+			 u8_t *rx_data, const int size);
+static void uart_nrf5e_fifo_rx_setup(struct device *dev);
+static void uart_nrf5e_pump_tx_fifo(struct device *dev);
+
+static void uart_nrf5e_mask(struct device *dev, int mask)
+{
+	volatile struct _uart *uart = UART_STRUCT(dev);
+	struct uart_nrf5e_dev_data_t * const dev_data = DEV_DATA(dev);
+
+	if (dev_data->in_interrupt) {
```

Comment:
Here, I am just wondering. Is it intended for the callers of this function to be invoked at an IRQ priority level that they could preemt the UARTE IRQ? If not, the in_interrupt could not be true, here, in my opinion.